pymongo/mongo_client.py

```diff
@@ -1397,7 +1397,14 @@ def is_retrying():
                             assert last_error is not None
                             raise last_error
                         retryable = False
-                    return func(session, sock_info, retryable)
+                    # Handle re-authentication.
+                    try:
+                        return func(session, sock_info, retryable)
+                    except OperationFailure as exc:
+                        if exc.code == helpers._REAUTHENTICATION_REQUIRED_CODE:
+                            sock_info.authenticate(reauthenticate=True)
+                            return func(session, sock_info, retryable)
+                        raise
```

Comment:
Is re-auth suppose to _only_ work with retryable read or write commands? Or is it supposed to work with _any_ command?

Reply:
It is supposed to work with any command, should it be moved into `sock_info.command`?

Reply:
There are essentially 3 places right now where we need this logic: 1) for cursors 2) for commands 3) for bulk writes

The 3 codepaths basically correspond to where session._apply_to is called:
```
git grep session._apply_to
pymongo/bulk.py:329:                    session._apply_to(cmd, retryable, ReadPreference.PRIMARY, sock_info)
pymongo/message.py:358:            session._apply_to(cmd, False, self.read_preference, sock_info)
pymongo/message.py:505:            self.session._apply_to(cmd, False, self.read_preference, sock_info)
pymongo/pool.py:760:            session._apply_to(spec, retryable_write, read_preference, self)
```

Reply:
Ideally we’d only have a single command dispatch code path but that may be difficult to wrangle. 

But yeah the retryable reads writes covers most (all?) of the cursor and bulk write code paths. Retrying within command path would probably cover the rest. We’d just need to make sure we don’t accidentally have two layers of retries.

Reply:
I updated the logic and added more tests.